examples/chardev2.c

```diff
@@ -136,12 +154,12 @@ device_ioctl(struct file *file, /* ditto */
          * be the device's message.  Get the parameter given to ioctl by
          * the process.
          */
-        temp = (char *)ioctl_param;
+        tmp = (char __user *)ioctl_param;
```

Comment:
Declare `tmp` here. That is,
```c
    case IOCTL_SET_MSG: {
        /* Receive a pointer to a message (in user space) and set that to
         * be the device's message.  Get the parameter given to ioctl by
         * the process.
         */
        char __user *tmp = (char __user *)ioctl_param;
        ...
    }
```

Reply:
It may lead Sparse report the error:
```shell
/lkmpg/examples/chardev2.c: In function ‘device_ioctl’:
/lkmpg/examples/chardev2.c:153:9: error: a label can only be part of a statement and a declaration is not a statement
  153 |         char __user *tmp = (char __user *)ioctl_param;
      |         ^~~~

```

Reply:
Oh, I found the example here: [linux/arch/alpha/kernel/smc37c669.c](https://github.com/torvalds/linux/blob/master/arch/alpha/kernel/smc37c669.c#L1233)
It seems fine. I will change it.

Reply:
It may lead the Cppcheck report error.
```shell
        <error id="variableScope" severity="style" msg="The scope of the variable &apos;ch&apos; can be reduced." verbose="The scope of the variable &apos;ch&apos; can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for &apos;i&apos; can be reduced:\012void f(int x)\012{\012    int i = 0;\012    if (x) {\012        // it&apos;s safe to move &apos;int i = 0;&apos; here\012        for (int n = 0; n &lt; 10; ++n) {\012            // it is possible but not safe to move &apos;int i = 0;&apos; here\012            do_something(&amp;i);\012        }\012    }\012}\012When you see this message it is always safe to reduce the variable scope 1 level." cwe="398">
            <location file="/home/slda/lkmpg/examples/chardev2.c" line="143" column="10"/>
            <symbol>ch</symbol>
        </error>
```

Reply:
We can safely suppress the warning "The scope of the variable ??? can be reduced" since it does not harm.

Reply:
> It may lead the Cppcheck report error.
> 
> ```shell
>         <error id="variableScope" severity="style" msg="The scope of the variable &apos;ch&apos; can be reduced." verbose="The scope of the variable &apos;ch&apos; can be reduced. Warning: Be careful when fixing this message, especially when there are inner loops. Here is an example where cppcheck will write that the scope for &apos;i&apos; can be reduced:\012void f(int x)\012{\012    int i = 0;\012    if (x) {\012        // it&apos;s safe to move &apos;int i = 0;&apos; here\012        for (int n = 0; n &lt; 10; ++n) {\012            // it is possible but not safe to move &apos;int i = 0;&apos; here\012            do_something(&amp;i);\012        }\012    }\012}\012When you see this message it is always safe to reduce the variable scope 1 level." cwe="398">
>             <location file="/lkmpg/examples/chardev2.c" line="143" column="10"/>
>             <symbol>ch</symbol>
>         </error>
> ```

Fixed by moving `char ch` into the scope.
```shell
-    char ch;
 
     /* Switch according to the ioctl called */
     switch (ioctl_num) {
@@ -151,6 +149,7 @@ device_ioctl(struct file *file, /* ditto */
          * the process.
          */
         char __user *tmp = (char __user *)ioctl_param;
+        char ch;
```